addons/mail/models/mail_template.py

```diff
@@ -191,126 +196,316 @@ def create_action(self):
+                    render_results=render_results,
+                    partners_only=partners_only
                 )
 
-            # Add report in attachments: generate once for all template_res_ids
-            if template.report_template:
-                for res_id in template_res_ids:
-                    attachments = []
-                    report_name = template._render_field('report_name', [res_id])[res_id]
-                    report = template.report_template
-                    report_service = report.report_name
+            # render scheduled_date
+            if 'scheduled_date' in render_fields:
+                template._generate_template_scheduled_date(
+                    template_res_ids,
+                    render_results=render_results
+            )
+
+            # add values static for all res_ids
+            template._generate_template_static(
+                template_res_ids,
+                render_fields,
+                render_results=render_results
+            )
 
-                    if report.report_type in ['qweb-html', 'qweb-pdf']:
-                        result, report_format = self.env['ir.actions.report']._render_qweb_pdf(report, [res_id])
-                    else:
-                        res = self.env['ir.actions.report']._render(report, [res_id])
-                        if not res:
-                            raise UserError(_('Unsupported report type %s found.', report.report_type))
-                        result, report_format = res
-
-                    # TODO in trunk, change return format to binary to match message_post expected format
-                    result = base64.b64encode(result)
-                    if not report_name:
-                        report_name = 'report.' + report_service
-                    ext = "." + report_format
-                    if not report_name.endswith(ext):
-                        report_name += ext
-                    attachments.append((report_name, result))
-                    results[res_id]['attachments'] = attachments
-
-            # hook for attachments-specific computation, used currently only for accounting
-            if 'attachments' in render_fields or 'attachment_ids' in render_fields and hasattr(self.env[self.model], '_process_attachments_for_template_post'):
-                records_attachments = self.env[self.model].browse(template_res_ids)._process_attachments_for_template_post(template)
-                for res_id, additional_attachments in records_attachments.items():
-                    if not additional_attachments:
-                        continue
-                    results[res_id]['attachment_ids'] += additional_attachments.get('attachment_ids', [])
-                    results[res_id]['attachments'] += additional_attachments.get('attachments', [])
+            # generate attachments if requested
+            if set(render_fields) & {'attachment_ids', 'report_template_ids'}:
```

Comment:
Seems you have a case for once creating a `render_fields_set` and using it on lines 479, 487, and 501.

Reply:
You really like sets don't you ?

Reply:
In this case you already used to sets. But indeed, where they're the best option, I do :-)
(Which might also be the case for the various checks in `_generate_template_static`, just saying ;))

Reply:
Yep I try to use them more often, still using list comprehensions a lot :D but I also like to keep input / output allowing any iterator as we don't have much clear prototyping currently ... hence sometimes I prefer to stay on list-like behavior and voila. Boarf.